### PR TITLE
Implementation of Threshold Activation Fn Done

### PR DESCRIPTION
**Pull request by shubham1206agra**
Related to #2181 

Reference: https://pytorch.org/docs/stable/generated/torch.nn.Threshold.html#torch.nn.Threshold

### CODE REVIEW

File: src/mlpack/methods/ann/layer/threshold_fn_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void Threshold<InputDataType, OutputDataType>::Forward(
+    const InputType& input, OutputType& output)
+{
+  output = arma::ones<OutputDataType>(arma::size(input));
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/threshold_fn_impl.hpp:36]: `set_size` should be sufficient here, no need to set the output to one first.

File: src/mlpack/methods/ann/layer/threshold_fn_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+template<typename DataType>
+void Threshold<InputDataType, OutputDataType>::Backward(
+    const DataType& input, const DataType& gy, DataType& g)
+{
+  DataType derivative;
+  derivative.set_size(arma::size(gy));
+  derivative.fill(0.0);
+  for (size_t i = 0; i < input.n_elem; ++i)
+  {
+    if (input(i) > threshold)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/threshold_fn_impl.hpp:53]: We have to use the same indices here as we used in the forward pass, maybe a good idea to store the indices for `input(i) > threshold` and use them here. Let me know if that sounds confusing.

File: src/mlpack/methods/ann/layer/threshold_fn_impl.hpp (hunk 1 above)
Comment (shubham1206agra) [reply src/mlpack/methods/ann/layer/threshold_fn_impl.hpp]: Done

File: src/mlpack/methods/ann/layer/threshold_fn_impl.hpp (hunk 2 above)
Comment (shubham1206agra) [reply src/mlpack/methods/ann/layer/threshold_fn_impl.hpp]: Done.

**Comment by shubham1206agra** [comment]
@zoq Can you now check it, please?
So that I can do another activation fn in the same branch only after merging.

File: src/mlpack/methods/ann/layer/threshold_fn.hpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the non zero gradient region threshold.
+  double const& Threshold_var() const { return threshold; }
+  //! Modify the non zero gradient region threshold.
+  double& Threshold_var() { return threshold; }
+
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/threshold_fn.hpp:93]: Could you change the style of the function's name to camel case?

File: src/mlpack/methods/ann/layer/threshold_fn.hpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+   * The non zero gradient region can be adjusted by specifying the parameter
+   * threshold. Default (alpha = 0.0)
+   * The value when x < threshold can be adjusted by specifying the parameter
+   * value. Default (alpha = 0.0)
+   *
+   * @param threshold Non zero gradient region threshold
+   * @param value Value when x < threshold
+   */
+  Threshold(const double threshold = 0.0,
+            const double value = 0.0);
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/threshold_fn.hpp:55]: Is value equal to alpha? If yes, would you change it to alpha in this case?

File: src/mlpack/methods/ann/layer/threshold_fn.hpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * @f{eqnarray*}{
+ * f(x) &=& \left\{
+ *   \begin{array}{lr}
+ *     x & : x > threshold \\
+ *     value & : x \le threshold
+ *   \end{array}
+ * \right.
+ * @f}
+ *
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/threshold_fn.hpp:31]: Is there any reference to this? Would you add the article that proposed this activation function as a reference?

File: src/mlpack/methods/ann/layer/threshold_fn.hpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class Threshold
+{
+ public:
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/threshold_fn.hpp:43]: I do not know about this, I understand that the activation function is called `Threshold` but maybe we need to change the class name? I am not sure about this, @zoq  any thought?

File: src/mlpack/tests/activation_functions_test.cpp [hunk 7]
```diff
@@ -751,6 +751,54 @@ TEST_CASE("ReLU6FunctionTest", "[ActivationFunctionsTest]")
...
+ * Implementation of the Threshold activation function derivative test. The function
+ * is implemented as Threshold layer in the file threshold_fn.hpp.
+ *
+ * @param input Input data used for evaluating the Threwshold activation function.
+ * @param target Target data used to evaluate the Threshold activation.
+ */
+void CheckThresholdCorrect(const arma::colvec input,
+                       const arma::colvec ActivationTarget,
+                       const arma::colvec DerivativeTarget)
+{
```
Comment (shrit) [inline src/mlpack/tests/activation_functions_test.cpp:764]: Would you fix the indentation her please?

File: src/mlpack/methods/ann/layer/threshold_fn.hpp (hunk 3 above)
Comment (shrit) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp:93]: You can have a more in depth look at mlpack style in the style guideline in the wiki.

**Review by shrit — COMMENTED** [review]
Look good, I left some comments, let me know when you resolve all these modifications.

File: src/mlpack/methods/ann/layer/threshold_fn.hpp (hunk 6 above)
Comment (shubham1206agra) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp:43]: Can you suggest me another name?

File: src/mlpack/methods/ann/layer/threshold_fn.hpp (hunk 4 above)
Comment (shubham1206agra) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp:55]: My bad. I left it by mistake.

File: src/mlpack/methods/ann/layer/threshold_fn.hpp (hunk 5 above)
Comment (shubham1206agra) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp:31]: I have only PyTorch docs as a reference.

File: src/mlpack/tests/activation_functions_test.cpp (hunk 7 above)
Comment (shubham1206agra) [reply src/mlpack/tests/activation_functions_test.cpp]: Done

File: src/mlpack/methods/ann/layer/threshold_fn.hpp (hunk 4 above)
Comment (shubham1206agra) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp]: Done.

File: src/mlpack/methods/ann/layer/threshold_fn.hpp (hunk 3 above)
Comment (shubham1206agra) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp]: Done.

File: src/mlpack/methods/ann/layer/threshold_fn.hpp (hunk 5 above)
Comment (shubham1206agra) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp:31]: @shrit

Comment (zoq) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp:31]: I don't think we will find a paper for the activation function, it#s super simple, so from my side this looks fine.

File: src/mlpack/methods/ann/layer/threshold_fn.hpp (hunk 6 above)
Comment (zoq) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp:43]: The name is fine for me.

File: src/mlpack/methods/ann/layer/threshold_fn.hpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
+/**
+ * @file methods/ann/layer/threshold_fn.hpp
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/threshold_fn.hpp:2]: To be consistent with the rest of the codebase, can you use `threshold.hpp`.

File: src/mlpack/methods/ann/layer/threshold_fn.hpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+ public:
+  /**
+   * Create the Threshold object using the specified parameters.
+   * The non zero gradient region can be adjusted by specifying the parameter
+   * threshold. Default (threshold = 0.0)
+   * The value when x < threshold can be adjusted by specifying the parameter
+   * value. Default (value = 0.0)
+   *
+   * @param threshold Non zero gradient region threshold
+   * @param value Value when x < threshold
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/threshold_fn.hpp:52]:
```suggestion
   * @param threshold Non zero gradient region threshold.
   * @param value Value when x < threshold.
```

See https://github.com/mlpack/mlpack/wiki/DesignGuidelines#comments for more information.

File: src/mlpack/methods/ann/layer/threshold_fn.hpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the non zero gradient region threshold.
+  double const& ThresholdVar() const { return threshold; }
+  //! Modify the non zero gradient region threshold.
+  double& ThresholdVar() { return threshold; }
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/threshold_fn.hpp:92]:
```suggestion
  double const& Threshold() const { return threshold; }
  //! Modify the non zero gradient region threshold.
  double& Threshold() { return threshold; }
```

File: src/mlpack/methods/ann/layer/threshold_fn.hpp [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+  OutputDataType outputParameter;
+
+  //! Locally-stored +ve output indexes object.
+  arma::uvec positive;
+
+  //! Threshold Parameter
+  double threshold;
+
+  //! Value Parameter
+  double value;
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/threshold_fn.hpp:122]:
```suggestion
  //! Threshold parameter.
  double threshold;

  //! Value parameter.
  double value;
```

File: src/mlpack/methods/ann/layer/threshold_fn_impl.hpp [hunk 12]
```diff
@@ -0,0 +1,68 @@
+/**
+ * @file methods/ann/layer/threshold_fn_impl.hpp
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/threshold_fn_impl.hpp:2]:
```suggestion
 * @file methods/ann/layer/threshold_impl.hpp
```

See comment above.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 13]
```diff
@@ -751,6 +751,54 @@ TEST_CASE("ReLU6FunctionTest", "[ActivationFunctionsTest]")
   CheckReLU6Correct(activationData, desiredActivations, desiredDerivatives);
 }
 
+/**
+ * Implementation of the Threshold activation function derivative test. The function
+ * is implemented as Threshold layer in the file threshold_fn.hpp.
```
Comment (zoq) [inline src/mlpack/tests/activation_functions_test.cpp:756]:
```suggestion
 * is implemented as Threshold layer in the file threshold.hpp.
```

File: src/mlpack/tests/activation_functions_test.cpp [hunk 14]
```diff
@@ -751,6 +751,54 @@ TEST_CASE("ReLU6FunctionTest", "[ActivationFunctionsTest]")
...
+ *
+ * @param input Input data used for evaluating the Threwshold activation function.
```
Comment (zoq) [inline src/mlpack/tests/activation_functions_test.cpp:758]:
```suggestion
 * @param input Input data used for evaluating the Threshold activation function.
```

File: src/mlpack/methods/ann/layer/threshold_fn.hpp (hunk 10 above)
Comment (shubham1206agra) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp:92]: Will it not conflict with class name

Comment (shubham1206agra) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp:92]: @zoq

Comment (zoq) [reply src/mlpack/methods/ann/layer/threshold_fn.hpp:92]: Ohh, you are right, in this case let's keep the method name.

File: src/mlpack/methods/ann/layer/threshold_impl.hpp [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void Threshold<InputDataType, OutputDataType>::Forward(
+    const InputType& input, OutputType& output)
+{
+  output.set_size(arma::size(input));
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/threshold_impl.hpp:36]:
```
output = input;
output(arma:find(input <= threshold)) = value
```
Just a thought,
Would something like the code above be better, we wouldn't storing an extra variable.

Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/threshold_impl.hpp:36]: Got it, thanks!
I just thought input was also being passed to backward and you can compute it there as well.

File: src/mlpack/methods/ann/layer/threshold_impl.hpp [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename DataType>
+void Threshold<InputDataType, OutputDataType>::Backward(
+    const DataType& /* input */, const DataType& gy, DataType& g)
+{
+  DataType derivative;
+  derivative.set_size(arma::size(gy));
+  derivative.fill(0.0);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/threshold_impl.hpp:49]: You can use `derivative.zeros(arma::size(gy));`.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 17]
```diff
@@ -751,6 +751,54 @@ TEST_CASE("ReLU6FunctionTest", "[ActivationFunctionsTest]")
...
+
+  // desiredActivations taken from PyTorch.
+  const arma::colvec desiredActivations("0.5 3.0 0.5 6.0 24.0 0.5");
+
+  // desiredDerivatives taken from PyTorch.
+  const arma::colvec desiredDerivatives("0.0 1.0 0.0 1.0 1.0 0.0");
+
+  CheckThresholdCorrect(activationData, desiredActivations, desiredDerivatives);
+}
+
```
Comment (zoq) [inline src/mlpack/tests/activation_functions_test.cpp:800]:
```suggestion
```

Remove the extra empty line.

**Pull request closed without merging**